### ISSUE TITLE
Keptn 0.14.1 multi-stage delivery broken as shipyard merges cloud event attributes of previous stages into current stage

### ISSUE THREAD

**Issue by christian-kreuzberger-dtx** [keptn/keptn#7627]
With Keptn 0.14.1, I ran into a problem with the sockshop example (any other example would lead to the same problem, the problem is in shippy, not in the example!).

**Problem**: Shipyard-controller is merging `deployment` data of previous stages into the current stage

## Dev Stage

`dev.delivery.triggered`:
```
{
  "data": {
    "configurationChange": {
      "values": {
        "image": "docker.io/keptnexamples/carts:0.13.1"
      }
    },
    "deployment": {
      "deploymentURIsLocal": null,
      "deploymentstrategy": ""
    },
    "project": "sockshop",
    "service": "carts",
    "stage": "dev"
  },
   ...
  "type": "sh.keptn.event.dev.delivery.triggered"
}
```

`deployment.triggered` by shipyard-controller in `dev` stage:
```
{
  "data": {
    "configurationChange": {
      "values": {
        "image": "docker.io/keptnexamples/carts:0.13.1"
      }
    },
    "deployment": {
      "deploymentURIsLocal": null,
      "deploymentstrategy": "direct"
    },
    "message": "",
    "project": "sockshop",
    "result": "",
    "service": "carts",
    "stage": "dev",
    "status": ""
  },
   ...
  "type": "sh.keptn.event.deployment.triggered"
}
```

`deployment.finished`  of helm-service in `dev` stage:
```
{
  "data": {
    "deployment": {
      "deploymentNames": [
        "direct"
      ],
      "deploymentURIsLocal": [
        "http://carts.sockshop-dev:80"                                              <--- OKAY
      ],
      "deploymentURIsPublic": [
        "http://carts.sockshop-dev.svc.cluster.local:80"                    <--- OKAY
      ],
      "deploymentstrategy": "direct",
      "gitCommit": "7f727d5ee12c0b1b69a630922211245f6b579435"
    },
    "message": "Successfully deployed",
    "project": "sockshop",
    "result": "pass",
    "service": "carts",
    "stage": "dev",
    "status": "succeeded"
  },
  ...
  "type": "sh.keptn.event.deployment.finished"
}
```

**Everything good in dev stage**

## Staging (2nd stage)

`staging.delivery.triggered`: 
```
{
  "data": {
    "configurationChange": {
      "values": {
        "image": "docker.io/keptnexamples/carts:0.13.1"
      }
    },
    "deployment": {
      "deploymentNames": [
        "direct"
      ],
      "deploymentURIsLocal":  [
        "http://carts.sockshop-dev:80"                                        <--- here we see the dev URL (this used to be null in 0.13.x and older)
      ],
      "deploymentURIsPublic": [
        "http://carts.sockshop-dev.svc.cluster.local:80"             <--- here we see the dev URL (this used to be here aswell in 0.13.x and older)
      ],
      "deploymentstrategy": "",
      "gitCommit": "7f727d5ee12c0b1b69a630922211245f6b579435"
    },
    "evaluation": {
      "gitCommit": "",
      "indicatorResults": null,
      "result": "pass",
      "score": 0,
      "sloFileContent": "",
      "timeEnd": "2022-05-03T06:33:48.991Z",
      "timeStart": "2022-05-03T06:33:45.898Z"
    },
    "message": "Finished release",
    "project": "sockshop",
    "release": {
      "gitCommit": ""
    },
    "result": "",
    "service": "carts",
    "stage": "staging",
    "status": "",
    "test": {
      "end": "2022-05-03T06:33:48.991Z",
      "gitCommit": "",
      "start": "2022-05-03T06:33:45.898Z"
    }
  },
   ...
  "type": "sh.keptn.event.staging.delivery.triggered"
}
```

**Observation**: `deployment.deploymentURIsLocal` and `deployment.deploymentURIsPublic`  are set to `"http://carts.sockshop-dev.svc.cluster.local:80"`, although we are in staging stage. I guess this is expected, as we merge the data from the previous stage.
However, we need to be careful with that, especially as the services in staging would not expect to find a dev URL here.

`deployment.triggered` by shipyard-controller in `staging` stage:

```
{
  "data": {
    "configurationChange": {
      "values": {
        "image": "docker.io/keptnexamples/carts:0.13.1"
      }
    },
    "deployment": {
      "deploymentNames": [
        "direct"
      ],
      "deploymentURIsLocal": [
        "http://carts.sockshop-dev:80"
      ],
      "deploymentURIsPublic": [
        "http://carts.sockshop-dev.svc.cluster.local:80"
      ],
      "deploymentstrategy": "blue_green_service",
      "gitCommit": "7f727d5ee12c0b1b69a630922211245f6b579435"
    },
    "evaluation": {
      "gitCommit": "",
      "indicatorResults": null,
      "result": "pass",
      "score": 0,
      "sloFileContent": "",
      "timeEnd": "2022-05-03T06:33:48.991Z",
      "timeStart": "2022-05-03T06:33:45.898Z"
    },
    "message": "",
    "project": "sockshop",
    "release": {
      "gitCommit": ""
    },
    "result": "",
    "service": "carts",
    "stage": "staging",
    "status": "",
    "test": {
      "end": "2022-05-03T06:33:48.991Z",
      "gitCommit": "",
      "start": "2022-05-03T06:33:45.898Z"
    }
  },
   ...
  "type": "sh.keptn.event.deployment.triggered"
}
```

**Observation**: We are providing very weird data from previous stage to helm-service here..., e.g., `test`, `release`, `evaluation`, and  `deployment`.
This might already confuse `helm-service` (or any other deployment service).

However, `helm-service` seems to cope fine with that, and respond with the correct `deployment.finished` event:
```
{
  "data": {
    "deployment": {
      "deploymentNames": [
        "canary"
      ],
      "deploymentURIsLocal": [
        "http://carts.sockshop-staging:80"                                      <--- OKAY
      ],
      "deploymentURIsPublic": [
        "http://carts.sockshop-staging.svc.cluster.local:80"            <--- OKAY
      ],
      "deploymentstrategy": "duplicate",
      "gitCommit": "c043d7f5817d3718b87eb40741a7ab0e794a0f44"
    },
    "message": "Successfully deployed",
    "project": "sockshop",
    "result": "pass",
    "service": "carts",
    "stage": "staging",
    "status": "succeeded"
  },
  ...
  "type": "sh.keptn.event.deployment.finished"
}
```

Now my staging stage has a `test` step, so shipyard-controller sends `test.triggered`:
```
{
  "data": {
    "configurationChange": {
      "values": {
        "image": "docker.io/keptnexamples/carts:0.13.1"
      }
    },
    "deployment": {
      "deploymentNames": [
        "direct",                                                                             <--- NOT OKAY
        "canary",                                                                           <--- OKAY
      ],
      "deploymentURIsLocal": [
        "http://carts.sockshop-dev:80",                                          <--- NOT OKAY
        "http://carts.sockshop-staging:80"                                     <--- OKAY
      ],
      "deploymentURIsPublic": [
        "http://carts.sockshop-dev.svc.cluster.local:80",                 <--- NOT OKAY
        "http://carts.sockshop-staging.svc.cluster.local:80"            <--- OKAY
      ],
      "deploymentstrategy": "duplicate",
      "gitCommit": "c043d7f5817d3718b87eb40741a7ab0e794a0f44"
    },
    "evaluation": {
      "gitCommit": "",
      "indicatorResults": null,
      "result": "pass",
      "score": 0,
      "sloFileContent": "",
      "timeEnd": "2022-05-03T06:33:48.991Z",
      "timeStart": "2022-05-03T06:33:45.898Z"
    },
    "message": "",
    "project": "sockshop",
    "release": {
      "gitCommit": ""
    },
    "result": "pass",
    "service": "carts",
    "stage": "staging",
    "status": "succeeded",
    "test": {
      "end": "2022-05-03T06:33:48.991Z",
      "gitCommit": "",
      "start": "2022-05-03T06:33:45.898Z",
      "teststrategy": "performance"
    }
  },
  ...
  "type": "sh.keptn.event.test.triggered"
}
```

**Observation 1**: Shipyard-controller seems to merge all deployment attributes (in 0.13.x and older, it didn't merge  `deploymentURIsLocal`)
**Observation2 **: jmeter-service seems to take `deploymentURIsLocal` at index 0, therefore already testing a service in the wrong stage (it should test *staging*) - it is testing `http://carts.sockshop-dev:80` instead of `http://carts.sockshop-staging:80` (Note: this is not a bug of jmeter-service, but shipyard-controller! jmeter-service should stay dumb, and only take a URL, and not figure out the URL itself).

## Production Stage

last but not least, this behaviour gets super weird in production stage:

`production.delivery.triggered`:

```
{
  "data": {
    "configurationChange": {
      "values": {
        "image": "docker.io/keptnexamples/carts:0.13.1"
      }
    },
    "deployment": {
      "deploymentNames": [                                      <--- WTF
        "direct",                                   
        "canary",
        "direct"
      ],
      "deploymentURIsLocal":  [  
        "http://carts.sockshop-dev:80",                  <--- WTF
        "http://carts.sockshop-staging:80",            <--- WTF
        "http://carts.sockshop-dev:80"                   <--- WTF
      ],
      "deploymentURIsPublic": [  
        "http://carts.sockshop-dev.svc.cluster.local:80",                  <--- WTF
        "http://carts.sockshop-staging.svc.cluster.local:80",            <--- WTF
        "http://carts.sockshop-dev.svc.cluster.local:80"                   <--- WTF
      ],
      "deploymentstrategy": "",
      "gitCommit": "7f727d5ee12c0b1b69a630922211245f6b579435"
    },
    "evaluation": {
      "gitCommit": "",
      "indicatorResults": null,
      "result": "pass",
      "score": 0,
      "sloFileContent": "",
      "timeEnd": "2022-05-03T06:33:48.991Z",
      "timeStart": "2022-05-03T06:33:45.898Z"
    },
    "message": "Finished release",
    "project": "sockshop",
    "release": {
      "gitCommit": ""
    },
    "result": "",
    "service": "carts",
    "stage": "production",
    "status": "",
    "test": {
      "end": "2022-05-03T06:33:48.991Z",
      "gitCommit": "",
      "start": "2022-05-03T06:33:45.898Z"
    }
  },
  ...
  "type": "sh.keptn.event.production.delivery.triggered"
}
```

# Summary

- Delivery use-case works with Keptn 0.12.x and 0.13.x, but it could already lead to weird behaviour because of the merging of attributes
- Attribute merging in 0.12.x and 0.13.x seems to have been done in another way than in 0.14.x
- Delivery use-case is broken in 0.14.x because jmeter-service will test against the wrong service (this is a symptom, the bug is in the merging of attributes/cloud events of shipyard-controller)

**Comment by christian-kreuzberger-dtx** [comment]
I verified that the same behaviour is also there for 0.12.x:

```
{
  "data": {
    "configurationChange": {
      "values": {
        "image": "docker.io/mongo:4.2.2"
      }
    },
    "deployment": {
      "deploymentNames": [
        "direct",
        "direct",
        "direct"
      ],
      "deploymentURIsLocal": null,                                                             <--- That would be correct
      "deploymentURIsPublic": [ 
        "http://carts-db.sockshop-dev.svc.cluster.local:80",                          <--- WTF
        "http://carts-db.sockshop-staging.svc.cluster.local:80",                    <--- WTF
        "http://carts-db.sockshop-dev.svc.cluster.local:80"                           <--- WTF
      ],
      "deploymentstrategy": "",
      "gitCommit": "c625ac7a2561d8124792d9ba2607c88e019cae86"
    },
    "message": "Finished release",
    "project": "sockshop",
    "release": {
      "gitCommit": ""
    },
    "result": "",
    "service": "carts-db",
    "stage": "production",
    "status": ""
  },
   ...
  "type": "sh.keptn.event.production.delivery-direct.triggered",
}
```

**The key difference** between 0.14.x and older versions is that `deploymentURIsLocal` is also merged in 0.14.x, while in older versions it was set to null.

**Comment by christian-kreuzberger-dtx** [comment]
This has been adressed with the image of PR #7631
